lib/pg.ddl.js

```diff
@@ -0,0 +1,395 @@
+  throw new Error(`Unsupported domen: ${domainName}`);
+};
+
+const generateTypes = domains => {
+  const sqlQueries = [];
+  const types = iter(domains)
+    .filter(([domainName]) => !IGNORED_DOMAINS.includes(domainName))
+    .map(([domainName, domain]) => {
+      const { type, sql } = generateType(domainName, domain);
+      if (sql) {
+        sqlQueries.push(sql);
+      }
+      return [domainName, type];
+    })
+    .collectTo(Map);
+  return { types, typesSQL: sqlQueries.join('\n\n') };
+};
+
+const requiresForeignKey = (categories, link, from) => {
+  const destinationType = categories[from].constructor.name;
+  return destinationType === 'Object' ||
+    HAS_FOREIGN_KEY.includes(destinationType);
+};
+
+const categorizeEntries = (category, categoryName, categories) => {
+  const indexes = [];
+  const unique = [];
+  const links = [];
+  const properties = [];
+  iter(Object.keys(category))
+    .filter(name => typeof category[name] !== 'function')
+    .each(name => {
+      if (!isValidIdentifier(name)) {
+        throw new Error(
+          `Cannot create entry ${categoryName}.${name} ` +
+            `because '${name}' is not a valid identifier`
+        );
+      }
+      if (isReserved(name)) {
+        throw new Error(
+          `Cannot create entry ${categoryName}.${name} ` +
+            `because '${name}' is a reserved keyword`
+        );
+      }
+      const property = category[name];
+      const type = property.constructor.name;
+      const entry = { name, property };
+      if (property.index) {
+        indexes.push(entry);
+      } else if (property.unique) {
+        unique.push(entry);
+      }
+      if (property.domain) {
+        properties.push(entry);
+      } else if (property.category) {
+        entry.foreignKey = requiresForeignKey(
+          categories,
+          property,
+          categoryName
+        );
+
+        if (type !== 'Many') {
+          if (type !== 'Object') {
+            entry.property.required = type === 'Master' || type === 'Include';
+          }
+
+          properties.push(entry);
+        }
+
+        if (type === 'Many' || entry.foreignKey) {
+          links.push(entry);
+        }
+      } else if (type === 'Index') {
+        indexes.push(entry);
+      } else if (type === 'Unique') {
+        unique.push(entry);
+      }
+    });
+
+  return { indexes, unique, links, properties };
+};
+
+const getMaxPropLength =
+  properties => iter(properties)
+    .map(property => property.name.length)
+    .reduce(Math.max);
+
+const generateProperties =
+  (category, properties, types, maxPropLength) => properties.map(
+    ({ name, property }) => {
+      const type = property.domain ? types.get(property.domain) : 'bigint';
+      if (!type) {
+        throw new Error(
+          `Domain ${property.domain} referenced from ${category}.${name} ` +
+          'is not defined'
+        );
+      }
+      let sql = `${pad(name, maxPropLength)} ${type}`;
+      if (property.required) {
```

Comment:
Nit: Can be inlined in the return statement.
`return !propetry.required ? sql : sql + ' NOT NULL';`